src/NamedTypesRegistry.cc

```diff
@@ -220,15 +343,59 @@ namespace drafter {
 
         FindNamedTypes(elementCollection, found);
 
-        std::sort(found.begin(), found.end(), InheritanceComparator(found));
+#ifdef DEBUG_DEPENDENCIES
+        std::cout << "==DEPENDENCIES INFO BEGIN==" << std::endl;
+#endif /* DEBUG_DEPENDENCIES */
+
+        DependencyTypeInfo typeInfo(found);
+
+        std::sort(found.begin(), found.end(), InheritanceComparator(typeInfo));
+
+#ifdef DEBUG_DEPENDENCIES
+        std::cout << "==BASE TYPE ORDER==" << std::endl;
+#endif /* DEBUG_DEPENDENCIES */
+
+        // first level registration - we will create empty elements with correct type info
+        for (DataStructures::const_iterator i = found.begin(); i != found.end(); ++i) {
+            const std::string& name = i->node->name.symbol.literal;
+
+            RefractElementFactory& factory = FactoryFromType(typeInfo.ResolveType(i->node));
+            refract::IElement* element = factory.Create(std::string(), false);
+            element->meta["id"] = refract::IElement::Create(name);
+
+            GetNamedTypesRegistry().add(element);
+        }
 
         for (DataStructures::const_iterator i = found.begin(); i != found.end(); ++i) {
 
             if (!i->node->name.symbol.literal.empty()) {
+
+                const std::string& name = i->node->name.symbol.literal;
                 refract::IElement* element = MSONToRefract(*i);
+#ifdef DEBUG_DEPENDENCIES
```

Comment:
Please make sure there is an empty line before and after these kind of blocks for easier reading comprehension.